File: src/main/javascript/crypto/src/libs/ERC721Token.ts
import {ASNEncodable} from "./ASNEncodable";
import {AsnParser} from "@peculiar/asn1-schema";
import {bnToUint8, uint8ToBn, uint8tohex} from "./utils";
import {ERC721} from "../asn1/shemas/NFTAttestation";
import {Asn1Der} from "./DerUtility";

export class ERC721Token implements ASNEncodable {
    private encoding: string;
    private address: string;
    private tokenId: bigint;

    static fromStrings(address:string, tokenId:string) {
        let me = new this();
        // remove 0x
        me.address = address.toLowerCase().substring(2);
        let tokenIdInteger: bigint;
        try {
            tokenIdInteger = BigInt(tokenId);
        } catch (e) {
            tokenIdInteger = 0n;
        }
        me.validateID(tokenIdInteger);
        me.tokenId = tokenIdInteger;
        me.encoding = me.constructEncoding();
        return me;
    }

    public fromStringAndBigint(address:string, tokenId:bigint) {
        this.address = address.toLowerCase();
        this.validateID(tokenId);
        this.tokenId = tokenId;
        this.encoding = this.constructEncoding();
    }

    private validateID(tokenId: bigint) {
        // Only allow non-negative IDs
        if (tokenId < 0n) {
            throw new Error("IDs cannot be negative");
        }
    }

    static fromBytes(asn1: Uint8Array) {
        try {
            let token = AsnParser.parse(asn1, ERC721);
            return ERC721Token.fromASNObj(token);
        } catch (e){
            throw new Error('Cant parse AttestationRequest Identifier');
        }
    }

    static fromASNObj(token: ERC721): ERC721Token {
        let me = new this();
        // Remove the # added by BouncyCastle
        //this.address = address.toString().substring(1);
        me.address = uint8tohex(new Uint8Array(token.address));
        me.tokenId = uint8ToBn(new Uint8Array(token.tokenId));
        me.encoding = me.constructEncoding();

        return me;
    }

    public getAddress():string {
        return this.address;
    }

    public getTokenId(): bigint {
        return this.tokenId;
    }

    public getDerEncoding(): string
    {
        return this.encoding;
    }

    public constructEncoding(): string {
        try {
            return Asn1Der.encode('SEQUENCE_30',this.getTokenVector());
        } catch (e) {
            throw new Error(e);
        }
    }
    public getTokenVector():string {
        return Asn1Der.encode('OCTET_STRING',this.address) +
            Asn1Der.encode('OCTET_STRING',uint8tohex(bnToUint8(this.tokenId)));
    }
}

File: src/main/java/io/alchemynft/attestation/ERC721Token.java
package io.alchemynft.attestation;

import java.io.IOException;
import java.math.BigInteger;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.bouncycastle.asn1.ASN1EncodableVector;
import org.bouncycastle.asn1.ASN1InputStream;
import org.bouncycastle.asn1.ASN1OctetString;
import org.bouncycastle.asn1.ASN1Sequence;
import org.bouncycastle.asn1.DEROctetString;
import org.bouncycastle.asn1.DERSequence;
import org.tokenscript.attestation.core.ASNEncodable;
import org.tokenscript.attestation.core.ExceptionUtil;
import org.web3j.utils.Numeric;

/**
 * Legacy class only used in Alchemy nft attestation. For any other use, use ERC721Token from
 * org.tokenscript.attestation
 */
@Deprecated
public class ERC721Token implements ASNEncodable {

  private static final Logger logger = LogManager.getLogger(ERC721Token.class);

  private final byte[] encoding;
  private final String address;
  private final BigInteger tokenId;

  public ERC721Token(String address, String tokenId) {
    this.address = address.toLowerCase();
    BigInteger tokenIdInteger;
    try {
      tokenIdInteger = new BigInteger(tokenId);
    } catch (Exception e) {
      tokenIdInteger = BigInteger.ZERO;
    }
    validateID(tokenIdInteger);
    this.tokenId = tokenIdInteger;
    this.encoding = constructEncoding();
  }

  public ERC721Token(String address, BigInteger tokenId) {
    this.address = address.toLowerCase();
    validateID(tokenId);
    this.tokenId = tokenId;
    this.encoding = constructEncoding();
  }

  public ERC721Token(byte[] derEncoding) throws IOException {
    ASN1InputStream input = null;
    try {
      input = new ASN1InputStream(derEncoding);
      ASN1Sequence asn1 = ASN1Sequence.getInstance(input.readObject());
      ASN1OctetString address = DEROctetString.getInstance(asn1.getObjectAt(0));
      ASN1OctetString tokenId = DEROctetString.getInstance(asn1.getObjectAt(1));
      // Remove the # added by BouncyCastle
      this.address = address.toString().substring(1);
      this.tokenId = new BigInteger(1, tokenId.getOctets());
      this.encoding = constructEncoding();
    } finally {
      input.close();
    }
  }

  private void validateID(BigInteger tokenId) {
    // Only allow non-negative IDs
    if (tokenId.compareTo(BigInteger.ZERO) < 0) {
      throw ExceptionUtil.throwException(logger,
          new IllegalArgumentException("IDs cannot be negative"));
    }
  }

  public String getAddress() {
    return address;
  }

  public BigInteger getTokenId() {
    return tokenId;
  }


  @Override
  public byte[] getDerEncoding() {
    return encoding;
  }

  byte[] constructEncoding() {
    ASN1EncodableVector data = getTokenVector();
    try {
      return new DERSequence(data).getEncoded();
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
  }

  public ASN1EncodableVector getTokenVector() {
    ASN1EncodableVector data = new ASN1EncodableVector();
    data.add(new DEROctetString(Numeric.hexStringToByteArray(address)));
    data.add(new DEROctetString(tokenId.toByteArray()));
    return data;
  }
}